### DISCUSSION TITLE
Complement of disjunction

### DISCUSSION THREAD

**Discussion by hansbogert** [cue-lang/cue#3229] [Q&A] [answered]
To preempt the A/B question I'll post a longer example, and my presumed minimal reproducer thereafter. 

I'm trying to do the following:

```cue
#systemProjectId: "someId"
#systemProjectNames: "system"|"monitoring"

#projects: [... ({
  name: #systemProjectNames
  annotations: systemProjectId: #systemProjectId
} | {
  name: !#systemProjectNames
})]


projects:  #projects  & [
  {
    name: "system"
    annotations: systemProjectId: "someId"
  },
  {
    name: "other"
  }
]
```
I get 

```
projects.1: 4 errors in empty disjunction:
projects.1.name: 2 errors in empty disjunction:
projects.1.name: conflicting values "monitoring" and "other":
    ./test.cue:2:31
    ./test.cue:4:13
    ./test.cue:5:9
    ./test.cue:12:12
    ./test.cue:18:11
projects.1.name: conflicting values "system" and "other":
    ./test.cue:2:22
    ./test.cue:4:13
    ./test.cue:5:9
    ./test.cue:12:12
    ./test.cue:18:11
projects.1.name: unresolved disjunction "system" | "monitoring" (type string):
    ./test.cue:8:9
```

I think the minimal reproducer is:

```
#foo: "1" | "2"

foo: !#foo & "3"
```

I get:

```
foo: unresolved disjunction "1" | "2" (type string):
    ./test2.cue:3:6
```

I'm not sure what unresolved means in this case. Conceptually this does not seem like much of a stretch to ask of the typing system, am I wrong?

**Comment by myitcv** [comment, marked as answer]
Hi @hansbogert - unfortunately you have just run into https://github.com/cue-lang/cue/issues/2656.

Your intuition around using `!` is solid: however in CUE that is actually spelt a different way. 

What you tried to write was ([play](https://cuelang.org/play/?id=5_A5nF5pVmA#w=function&i=cue&f=export&o=cue)):

```cue
#foo: "1" | "2"

foo: !#foo & "3"
```

In CUE today that needs to be written as follows ([play](https://cuelang.org/play/?id=9kX37kk0spN#w=function&i=cue&f=export&o=cue)):

```cue
#foo: "1" | "2"

foo: "3"
_foo_ok: true & ((foo & #foo) == _|_)
```

In future (#943) this will be written:

```cue
#foo: "1" | "2"

foo: not(#foo) & "3"
```

**Reply by hansbogert** [reply]
Thank you for the concise answer. Hoping for the `not()` in the future, because that looks neat.

**Reply by myitcv** [reply]
@jpluscplusm - a good hopefully short-lived doc here.

**Reply by jpluscplusm** [reply]
https://github.com/cue-lang/docs-and-content/issues/160 opened.

**Reply by jpluscplusm** [reply]
@hansbogert was kind enough to write a howto guide on this topic. It's been published at https://cuelang.org/docs/howto/negate-a-disjunction/

Thank you, Hans! 🙌
